admin/code/CMSForm.php

```diff
@@ -8,26 +8,34 @@
  * @subpackage admin
  */
 class CMSForm extends Form {
-	
+
+	protected $responseNegotiator;
+
 	/**
 	 * Route validation error responses through response negotiator,
 	 * so they return the correct markup as expected by the requesting client.
 	 */
-	protected function getValidationErrorResponse() {
+	protected function getValidationErrorResponse(ValidationResult $result) {
 		$request = $this->getRequest();
 		$negotiator = $this->getResponseNegotiator();
 
 		if($request->isAjax() && $negotiator) {
-			$this->setupFormErrors();
+			// Load form errors from the result into the form
+			// Also save them to session, in case the negotation returns a 302
+			$this->setupFormErrors($result, $this->getData());
+			$this->saveFormErrorsToSession($result, $this->getData());
+
 			$result = $this->forTemplate();
```

Comment:
Careful; Now $result is re-used with two different types. I would expect to see a new variable used for the actual response body.